### ISSUE TITLE
Tracking Issue for perma-unstable `wasm-c-abi` flag

### ISSUE THREAD

**Issue by daxpedda** [rust-lang/rust#122532]
This is a tracking issue for the MCP "perma-unstable `wasm-c-abi` flag" (https://github.com/rust-lang/compiler-team/issues/703).

For background, see https://github.com/rust-lang/compiler-team/issues/703. The short summary is that the ABI rustc uses for `extern "C"` on wasm32-unknown-unknown is not spec-compliant and exposes various unstable implementation details of rustc and LLVM. We plan to transition to the official "C" ABI for this target. This will change how non-scalar arguments are passed across this ABI boundary, so if you have code running that is built for wasm32-unknown-unknown and called directly by JavaScript, it may have to be adjusted to follow the new ABI rules. Options for how to manage the transition are still being considered; see https://github.com/rust-lang/rust/issues/138762 for further guidance.

The `-Zwasm-c-abi` flag lets you test your code with the new ABI.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Implement the MCP (cc @rust-lang/compiler -- can anyone write up mentoring
      instructions?)
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- ~~Formatting for new syntax has been added to the [Style Guide] ([nightly-style-procedure])~~
- ~~Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])~~

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs
[nightly-style-procedure]: https://github.com/rust-lang/style-team/blob/master/nightly-style-procedure.md 
[Style Guide]: https://github.com/rust-lang/rust/tree/master/src/doc/style-guide

### Implementation history

- https://github.com/rust-lang/rust/pull/117919
- https://github.com/rust-lang/rust/pull/117918
- https://github.com/rust-lang/rust/pull/133951
- https://github.com/rust-lang/rust/pull/138601

### Related issues
- https://github.com/rust-lang/rust/issues/115666
- https://github.com/rust-lang/rust/issues/138762

**Comment by workingjubilee** [comment]
Since 1.78 a future-compat-warning for using too-old versions of wasm-bindgen has been in place, and now 75% of the ecosystem's downloads of crates are wasm-bindgen 0.2.89 or higher. I have opened #129534 as a discussion point for whether we should move the warning to a deny.

**Comment by lcnr** [comment]
given #129534

r? @workingjubilee @alexcrichton

does this need a compiler FCP?

**Comment by Manishearth** [comment]
Hi! I was recently informed that there are plans to switch the default soon.

I work on [Diplomat](https://github.com/rust-diplomat/diplomat), a tool that generates multi-language bindings (including JS-wasm), used by [ICU4X](https://github.com/unicode-org/icu4x),

My _expectation_ here was that there would be a period of time where _both_ compilation modes are available on stable before things are switched over, as is the norm for almost all breaking changes in Rust. I have been waiting for that period of time to migrate Diplomat, so that Diplomat (and ICU4X) users will have an "apply `RUSTFLAGS=foo`" path available to them to deal with this across supported Rust versions.

I'm rather surprised to hear that it's just going to be a simple switchover. It's not a _huge_ deal for Diplomat-WASM to require nightly for working around this, but it's a bit messy, and like I said it is surprising. (I'm also surprised that the extent of future-incompat work was to ensure people use the latest wasm-bindgen: wasm-bindgen isn't the only tool around, and there are people who work directly with WASM with no generated bindings!)

**Comment by hanna-kruppe** [comment]
The flag that this tracking issue is about has been described as “perma-unstable” ever since it was first proposed in https://github.com/rust-lang/compiler-team/issues/703

wasm-bindgen has been made compatible with the fixed, spec-compliant ABI a while ago - the only migration users of wasm-bindgen need is updating their project to a wasm-bindgen version that has the fix, which they can and should do before rustc changes the ABI. The unstable flag is only intended for testing that it actually works with the new ABI (if it doesn’t, that would likely be a bug in wasm-bindgen). Is there a reason why a more complicated migration would be needed for Diplomat?

**Comment by RalfJung** [comment]
Specifically, the PR that made that change in wasm-bindgen is https://github.com/rustwasm/wasm-bindgen/pull/3595. I don't fully understand what this does, but it looks like it restricts `extern "C"` functions to use types that behave the same with the old, buggy and the new, correct ABI. Other projects such as Diplomat should be able to follow the same approach. Unfortunately I do not know what exactly the set of "safe" types are where the existing ABI is correct.

**Comment by Manishearth** [comment]
> The flag that this tracking issue is about has been described as “perma-unstable” ever since it was first proposed in [rust-lang/compiler-team#703](https://github.com/rust-lang/compiler-team/issues/703)

Right. I imagined this meant that the actual cutover would be handled via some other mechanism, with this perma-unstable flag being the way things are experimented with. Not uncommon in cases like this, to implement a feature some perma-unstable way, and then eventually figuring out the stable interface. I imagined one stable interface might have been doing it via function annotations, which is messy but still migrateable/cfgable.

Like. This is a _breaking change_. Rust is careful about breaking changes. There are a lot of assumptions Rust users are used to making about breaking changes, and the route taken here breaks many of them.

The fact that this issue is entitled perma-unstable made it even _more_ of an expectation that the cutover was far away, because I expected the timeline to be: new ABI implemented -> someone RFCs a stabilizable mechanism -> stabilizable mechanism implemented -> stabilized -> default switched -> (removal of old way, cleanup).

Hell, a typical expectation is that you do not need to know about random Rust issue titles to know that a breaking change is coming, too, but I did actually know this was happening eventually, just not so soon.




> Specifically, the PR that made that change in wasm-bindgen is [rustwasm/wasm-bindgen#3595](https://github.com/rustwasm/wasm-bindgen/pull/3595). I don't fully understand what this does, but it looks like it restricts `extern "C"` functions to use types that behave the same with the old, buggy and the new, correct ABI. Other projects such as Diplomat should be able to follow the same approach. Unfortunately I do not know what exactly the set of "safe" types is where the existing ABI is correct.

No, what it does is effectively building up its own manual ABI that is scalar-splatted. There is no way to have code with nontrivial `extern "C"` types work the same with the new and old ABI here. The set of "safe" types here is primitives.


Diplomat is not a wasm FFI library. It is a multi language FFI library, which has JS-wasm as a single backend. We cannot just make `extern "C"`-interface decisions for WASM without breaking every other language. We are not wasm-bindgen and have different tradeoffs to deal with compared to wasm-bindgen. In theory we could potentially start generating two extern "C" functions; one with `cfg(target = wasm32-unknown-unknown)` or whatever, but that's pretty messy and still needs an entire pile of complicated splatting trait infrastructure. _Plus_ it breaks the case of linking to Diplomat via C++ in wasm/emscripten, which is a case we care about as well[^2]!


It also restricts itself to types that contain a maximum of 4 scalars. That is an arbitrary limit that can be increased in a messy way, but ICU4X already has types that are larger than that and Diplomat doesn't want to have a limit here. Diplomat would have to do something like this for the same reason wasm-bindgen does: Diplomat's macro, which generates `extern "C"` functions, does not necessarily know anything about the actual type being passed down, because the type may come from a different module. So it hardcodes a number of associated types into its trait infrastructure and then unconditionally treats each parameter as having 4 scalars, with the type system using `()` for unused scalars (apparently this works fine in both wasm ABIs: these parameters are ignored entirely).

wasm-bindgen's fix is actually somewhat intriguing: one thing I do like about the legacy ABI[^1] is that parameters don't need any Rust-heap allocations to be passed through. I prefer that! I'm probably going to have to implement a mini-allocator to improve performance with the new ABI. Would be good to have a way to get "always splat into primitives" as the ABI, perhaps via an attribute, but that's not something _necessary_ here, just something that might be worth designing at some point.

It seems like you probably want undecorated Rust `extern "C"` functions to follow the specced ABI, but it's less important for stuff handled by tooling, which is precisely where "apply an attribute to each function" actually works.


 [^1]: which I overall dislike strongly, it is extremely quirky in a way that is undocumented, I believe I have written the [only actual docs on this subject](https://github.com/rust-diplomat/diplomat/blob/main/docs/wasm_abi_quirks.md) and took _a while_ to discover all the quirks. I discovered a new quirk just this week, which is part of why I was looking into this issue at all.
 [^2]: ICU4X cared enough that we [used to CI it](https://github.com/unicode-org/icu4x/pull/968), though we stopped doing that at some point because it increased the number of dependencies needed for a full test build and the test was [effectively just testing emscripten](https://github.com/unicode-org/icu4x/pull/4142)

**Comment by hanna-kruppe** [comment]
> In theory we could potentially start generating two extern "C" functions; one with `cfg(target = wasm32-unknown-unknown)` or whatever, but that's pretty messy and still needs an entire pile of complicated splatting trait infrastructure. Plus it breaks the case of linking to Diplomat via C++ in wasm/emscripten, which is a case we care about as well!

I don’t understand how it would break emscripten interop? wasm32-unknown-emscripten is a separate target from wasm32-unknown-unknown, and only the latter has the broken non-standard "C" ABI that needs an arduous migration. The emscripten target may have been affected by the same ABI incompatibility when it was first introduced, but if so, that was unceremoniously fixed as a bug years ago. (I’ve not seen the emscripten target mentioned explicitly in https://github.com/rust-lang/rust/issues/71871 but I wouldn’t be surprised if it affected all wasm targets at the time.)

**Comment by bjorn3** [comment]
https://github.com/rust-lang/rust/pull/133951 (will be released in rustc 1.86) will make the correct C ABI for wasm the default and only stable option. I plan to remove `-Zwasm-c-abi` entirely once it has hit stable and no significant breakage is shown.

> No, what it does is effectively building up its own manual ABI that is scalar-splatted. There is no way to have code with nontrivial extern "C" types work the same with the new and old ABI here. The set of "safe" types here is primitives.

Yes there is. Any non-trivial types on wasm are passed by-reference, so just a pointer as argument will have the exact same ABI between the old and new ABI and match what the new ABI would be if you directly passed the non-trivial type. And for return values you did pass a pointer to write the return value to as first argument.

> Would be good to have a way to get "always splat into primitives" as the ABI

The old wasm ABI splats the LLVM type. This is ill defined around enums, unions and structs with padding however. And even for the other cases the LLVM type is an internal implementation detail that may change. The list of quirks you documented in https://github.com/rust-diplomat/diplomat/blob/main/docs/wasm_abi_quirks.md is precisely because the LLVM type was never meant to be exposed to the user in any way.

**Comment by bjorn3** [comment]
> The ABI exists entirely by accident. My understanding is that someone didn't realize they have to add some code in rustc's ABI handling whenever a new architecture is being added, and so the ABI is some undocumented artifact of what LLVM happens to do.

To be precise it is the undocumented artifact of the combination of what LLVM type rustc happened to generate and what LLVM happens to do when it sees a type impossible to represent at the wasm level.

**Comment by RalfJung** [comment]
> https://github.com/rust-lang/rust/pull/133951 (will be released in rustc 1.86) will make the correct C ABI for wasm the default and only stable option.

That PR does not seem to change the ABI? I am confused. The description says it just changes a lint level.

**Comment by bjorn3** [comment]
Right, got confused by which PR did the flip. The PR that does this hasn't been merged yet as it was first waiting on this PR to get to stable.

**Comment by RalfJung** [comment]
It seems what @Manishearth is saying is that Diplomat actually relies on the exact what the broken ABI happens to look. I don't quite understand how this is possible in a multi-language project since no other language implements that ABI, so I think we need some more details here.

**Comment by RalfJung** [comment]
> Hell, a typical expectation is that you do not need to know about random Rust issue titles to know that a breaking change is coming, too, but I did actually know this was happening eventually, just not so soon.

There's no way to tell when someone relies on the accidental ABI.

It might make sense to have a blog post about this to warn people that this is coming; I am a bit surprised that I couldn't find one.

**Comment by workingjubilee** [comment]
We will *not* support the ad-hoc "splatting ABI" on a *permanent* basis, because the degree of design that went into it makes the 32-bit x86 GNU/Windows ABI look *well-considered, carefully planned, and bug-free*. It is frankly ludicrous that people waffled around long enough for anyone to develop a dependence on its broken state, and we should not do that sort of thing anymore.

However I am willing to entertain stabilizing a flag that enables the "version 1" ABI per https://github.com/WebAssembly/tool-conventions/blob/main/BasicCABI.md now that we have [target modifiers](https://github.com/rust-lang/rust/issues/136966). It will, of course, do nothing in short order. It may be useful to have around in case we ever get a v2 of the wasm ABI.

**Comment by hanna-kruppe** [comment]
An opt-in target modifier is of little use in this situation because enabling it requires `-Zbuild-std` to avoid mismatch with the pre-built sysroot. If you're willing to use nightly for that you might as well use `-Zwasm-c-abi` as long as it exists -- and once it no longer exists, the target modifier for "v1 ABI" would also be turned on by default, so it no longer serves any purpose. It may be useful in the future for picking another well-defined ABI (e.g., https://github.com/WebAssembly/tool-conventions/issues/247) as the default, but that seems entirely independent of the current mess so I wouldn't consider that a reason to introduce the "select wasm ABI version" flag *now*.

**Comment by workingjubilee** [comment]
Ah yes, prebuilt std strikes again!

**Comment by Manishearth** [comment]
> I'm not quite sure what to reply to this... clearly you imagined a lot but unfortunately waited extremely long before raising any of these points. The transition plan has been communicated fairly clearly I think, 


Communicated _to whom_? The only reason I discovered this whole thing was that I was researching the weird ABI and _eventually_ I stumbled across this issue. There was no clear cut transition plan when I did stumble across this.


> and there's little we can do to prevent people from imagining other things that have never been alluded to even in the slightest.


No, _this is a Rust project norm around breaking changes_. I am willing to accept that the compiler team did not fully understand the impact of this change and treated this as "yet another soundness fix that few need to pay attention to". That's a mistake, but quite understandable. There _is_ ultimately no guarantee that you will always be able to understand the full set of people relying on a thing. But you can try.

However, I do not think I am incorrect in assuming that Rust norms will be followed. Saying that this is "imagined" on my behalf hurts Rust's hard-earned reputation on breaking changes. I don't think I am being unreasonable here.

> Anyone who uses `extern "C"` to interface with other languages will currently experience extremely broken behavior

Here's the thing: it is broken, period, for interfacing with non-JS languages. This means there are no current users there, only potential users. They are not being _fixed_, they are being _enabled_.


It is _confusing_ but workable for interfacing with JS. It's not broken in an unsound way, it works just fine, and Rust has touted "Rust + wasm32-unknown-unknown + JS" as a favored pairing for doing Wasm for six years now. I do not think it is at all unreasonable to expect some modicum of stability there.

The document lists out a confusing ABI, but .... honestly, the padding thing is the only actually weird bit. The rest of it is not surprising at all in the wasm ecosystem: wasm's non-multival handling of multiple scalars in the _return_ values is similarly confusing (and underdocumented), which brings me to this point:

> Next time before you reverse engineer rustc's output and then assume that somehow what you learn that way is in any way even remotely a stable promise, please ask for clarification first. :) Undocumented implementation details are generally never stable guarantees.

As mentioned, the _standard_ Wasm ABI for this is (or was?) also confusing, underdocumented, and in a state of "we're building a better thing, please hold" (multival). So the expectations here are a bit different I think.

Rust has been touting wasm as a strength for ages, and this "implementation detail" is _a thing you have to figure out to do some of the most basic tasks in rust-wasm_.

To be clear, since I'm not sure if the compiler team properly understands this from the comments I'm getting at here, this is something you need to figure out to do the task of "call an `extern "C"` Rust function from JS in Wasm, which takes an `repr(C)` aggregate parameter". This is not a complex task. This is not a niche task. **This is what I would consider a very normal FFI task**. If a toolchain says it supports FFI between JS and Rust, I would imagine that this is stable: JS code I write to interface with the Rust code doesn't need to change.

That's how Diplomat relies on this: we generate JS function calls like `wasm.some_extern_c_function(arg, arg, padding, arg, arg, ...)`. Presumably this is what wasm-bindgen did as well before their fix of manual-splatting.

Like yes, I will be more careful about  this in the future, but I wish to  still express extreme surprise that this was considered to be something that had no users that mattered.

Also, I _did_ ask around when I first discovered quirks here, I was pointed at the BasicCABI doc (which doesn't cover this), and nobody seemed to know much about Rust's behavior. I should have asked around more, clearly, but the wasm world doesn't have a good enough reputation of documenting things for me to have high expectations here.

> EDIT: Or maybe we could lint against passing any non-scalar type across that ABI? That lint should not fire for wasm-bindgen any more if I understand their fix properly.



Yes, this would have helped discover users.

> It might make sense to have a blog post about this to warn people that this is coming; I am a bit surprised that I couldn't find one.

So would this. I'd think it basically necessary for a change like this that breaks direct FFI for the primary language associated with FFI for a well-supported and advertised language.

> It seems what [@Manishearth](https://github.com/Manishearth) is saying is that Diplomat actually relies on the exact way the broken ABI happens to look, with consumers outside their control, or so? I don't quite understand how this is possible in a multi-language project since no other language implements that ABI, so I think we need some more details here.

We are a multi language project in that we generate _one_ set of `extern "C"` bindings that can be called from multiple languages. So for `impl SomeType { fn foo(x: Foo) {..}}`, where `Foo` is some `repr(C)` aggregate, we'll generate `extern "C" fn foo(x: Foo)`, similar to what wasm-bindgen used to do. Then, in C we'll generate `void foo(Foo)`. In C++ we'll generate a class method on `SomeType` that internally calls a `void foo(Foo)`, performing necessary conversions. In JS we'll generate a class method on `SomeType` that calls `wasm.foo(field1, padding1, field2, field3, ...)` with the exact call computed based on the quirks doc I've written.

Basically, JS — the primary, original main consumer of Wasm! — is usable with that ABI. Confusing yes, but usable. And confusing isn't that out of the norm for Wasm.


What is currently unusable is calling that ABI from e.g. C++ in wasm32-unknown-unknown. Which I haven't even tried in ages: the C++ world seems to universally favor emscripten in their wasm docs, and wasm32-unknown-unknown at least used to be broken for Rust-C++ in other ways (I don't think it even would link correctly?). I don't know the status quo, I internalized the lack of unknown-unknown docs, and the brokenness of the toolchain as "don't use unknown-uknown with C++" and moved on years ago.


@workingjubilee 

> We will _not_ support the ad-hoc "splatting ABI" on a _permanent_ basis, because the degree of design that went into it makes the 32-bit x86 GNU/Windows ABI look _well-considered, carefully planned, and bug-free_. It is frankly ludicrous that people waffled around long enough for anyone to develop a dependence on its broken state, and we should not do that sort of thing anymore.

Yeah, 100%. The more I learn about this the more I think it was a mistake to start the marketing push for Rust wasm back when it happened with the ABI being broken for some of the most basic tasks. Rust papered over it with wasm-bindgen as an official tool, so it took some time for people to actually do manual FFI, but there are plenty of reasons to do manual FFI and in general it is surprising if doing things "vanilla" is unsupported.

I think a thing Diplomat may be able to do is write a macro that performs the transform that wasm-bindgen does. It's not great, and it would need further trait infra, but it's ... fine.



@hanna-kruppe 

> I don’t understand how it would break emscripten interop? wasm32-unknown-emscripten is a separate target from wasm32-unknown-unknown, and only the latter has the broken non-standard "C" ABI that needs an arduous migration.

Yeah, ignore that comment. Diplomat tweaking the API to do the wasm-bindgen trick on unknown-unknown is possible and does not have that downside with this in mind. It is a lot of work though and we'd probably prefer to just do the new thing. Diplomat needs to discuss this internally, like I said I didn't know this was coming so soon.




---------------


All in all, I'm not really asking for anything here at this point (though maybe, a blog post? focusing on the users doing non-wasm-bindgen assisted JS-Rust FFI). I understand that this has been pretty broken for a while, and there's a strong desire to fix this. I understand that the compiler team may not have had a good understanding of the users around.

At this point, Diplomat will likely take the route of just requiring certain Rust versions or nightly flags. It's not ideal, it's fine, I don't actually wish to _slow down_ this process. It's taken long enough already, I don't wish to make it take another year to handle it "properly", and this is something I can suboptimally paper over on my side.

I mostly wish to express surprise that this happened, and wish for the compiler team to potentially introspect _how_ this came to happen and avoid it happening again. This has the potential to be a nontrivial reputational blow to Rust's stability guarantees.

For example, in an unsafe context, a lot of people have been making similar assumptions about tree borrows / provenance / etc: The stable APIs that enable tricky pointery use cases will be available well before the compiler starts exploiting the UB defined there. It's pretty standard in the unsafe review space, because there are lots of patterns which do not have a "good answer" yet. I know that provenance stuff got recently stabilized in a way that smoothly includes existing users (nice), but there's still more stuff in this space being crystallized.

This happening makes me worried about those assumptions as well, and makes me think that I need to be far more involved in paying attention to the changes in opsem and new compiler optimizations than I currently am. I ... don't think that is actually what the compiler team wishes its users to have to do.

Another thing I'd like to highlight: no shade on the maintainers of wasm-bindgen here[^1], but this issue being driven by wasm-bindgen maintainers and having a pervasive assumption that wasm-bindgen is the only use case that matters is a _really bad look_ for the Rust project. I personally don't care about ecosystem "competition" or whatever, but even just the _fear_ of the Rust project favoring particular ecosystem participants[^2] has caused plenty of drama in the past. This is something teams should be far more careful about. It's totally fine and even expected for ecosystem maintainers to be heavily involved in such issues, but you do have to be careful other participants don't get left behind.




 [^1]: I am very very glad for their work here! Thank you!
 [^2]: wasm-bindgen is in a weird place of being a former "official working group" project and currently in an unclear state of officialness, but I don't actually think officialness is relevant here

**Comment by bjorn3** [comment]
> That's how Diplomat relies on this: we generate JS function calls like wasm.some_extern_c_function(arg, arg, padding, arg, arg, ...). Presumably this is what wasm-bindgen did as well before their fix of manual-splatting.

Wasm-bindgen never used types that contained padding. It only passed simple scalars (unchanged) and structs consisting of up to 4 simple scalars (or unions of such structs with the unit type, which happens to be represented the same as the non-unit field). In fact there was a comment directly warning against any types containing padding: https://github.com/rustwasm/wasm-bindgen/pull/3595/files#diff-44d1a04f594a2cfc0cb0a831e7eee00b729c6b1c5427d85907ed632810fa262fL320-L341

**Comment by Manishearth** [comment]
Ah, I see. Wasm-bindgen handled this by always allocating user structs on the heap.

I was rather confused since "4 scalar fields max" is super super easy to hit (that's a struct with two option fields). But it seems like wasm-bindgen only needs to deal with this for _its own types_ (options, slices, etc), and not user types, because user types are unconditionally passed indirectly.

wasm-bindgen calls the IntoWasmAbi trait for _every single_ parameter so the macro does not need to know the types of anything. This ends up with some really ugly looking Rust function signatures[^1], but people aren't looking at those directly, so that's fine.

This is technically doable in Diplomat. But it's a whole bunch of trait infrastructure that we have not yet needed, so I think we'll probably just go the route of doing our own cutover at some point and requesting people use the right version of Rust or pass the unstable flag.


With that in mind, I would like to request that the unstable legacy ABI be kept around for a while: it's going to take some time for ICU4X's MSRV to get to the point where we can switch.

 [^1]: Basically it generates `arg0_1: <<Foo as FromWasmAbi>::Abi as WasmAbi>::Prim1, arg0_2: ...Prim2, ... arg1_0: <<Bar as ...>>::Prim1, ...` as the argument list. Each argument gets splatted into four.

**Comment by hanna-kruppe** [comment]
> For example, in an unsafe context, a lot of people have been making similar assumptions about tree borrows / provenance / etc: The stable APIs that enable tricky pointery use cases will be available well before the compiler starts exploiting the UB defined there. It's pretty standard in the unsafe review space, because there are lots of patterns which do not have a "good answer" yet. I know that provenance stuff got recently stabilized in a way that smoothly includes existing users (nice), but there's still more stuff in this space being crystallized.
>
> This happening makes me worried about those assumptions as well, and makes me think that I need to be far more involved in paying attention to the changes in opsem and new compiler optimizations than I currently am. I ... don't think that is actually what the compiler team wishes its users to have to do.

Note a significant difference between new APIs that are intended to eventually become "the right way" and ABI changes: new APIs can just be added next to existing ones at any time, while changing the interpretation of an ABI is inherently something of a hard cut. Making the change depend on a compiler flag instead of `rustup update` hardly helps: linking code compiled with one set of ABI-affecting flags to code compiled with a different set is a gaping soundness hole even for pure Rust code bases without any FFI to other languages. Thus, users still have to migrate entire projects at once, and get free bonus bugs if they don't do that. While Rust has (very recently!) gained the "target modifier" concept for managing such mismatches, that doesn't help stable users as long as `-Zbuild-std` remains unstable. These problems also affect `-Zwasm-c-abi`, which is only okay-ish because of four mitigating factors:

1. On wasm32-unknown-unknown, the precompiled sysroot generally doesn't talk to the rest of the Rust world in terms of `extern "C"` functions. Note that this is pretty fragile w.r.t. on internal implementation choices of the standard library. If there's any `extern "C"` symbols or function pointers used internally for any reason, we must either hope that they either fall into the subset of signatures where `-Zwasm-c-abi` doesn't make a difference, or that calls and definitions are never codegen'd under different `RUSTFLAG`s (i.e., are actually precompiled and not codegen'd in other crates later).
2. Iinterop with C code (or any other language) compiled to wasm has been so hopelessly broken for so long that nobody does it, so if all Rust code that does `extern "C"` calls is compiled with the same flag, the mutability of that ABI only matters for interop with host code.
3. For host code, the focus was on wasm-bindgen when considering if this flag should be added, and the whole point of the flag is to support wasm-bindgen in testing the work that had already been done to be compatible with the standard ABI. In contrast, I expect that a MCP along the lines of "there's 1000 small projects that do hand-rolled JS-RustWasm FFI and need to migrate, let's give them this flag so they can manage the migration on their own terms" would have been more controversial.
4. There's little if any dynamic linking between wasm modules in typical usage of wasm32-unknown-unknown. Directly connecting a function exported by one wasm module to an import of another wasm module is rare (or at least has been so far) so you won't usually run into issues due to separately-compiled Rust code meeting over an `extern "C"` boundary whose ABI they disagree about.

There's unfortunately no nice, gradual migration story for ABI changes. Every project with FFI touching on the (relevant parts of) the ABI has to either do a hard switch from supporting only the former interpretation of the ABI to only supporting the new one, or they have to restrict themselves for the common subset that works the same before and after. Since that's often quite difficult, ABI bug fixes (including on wasm targets!) have generally been treated as "ripping off the band aid" and not as a migration period.

The one exceptional aspect about wasm-bindgen is that investing migration effort into this one binding generator can absorb the pain of migration for a whole lot of users at once. But that calculus only applies to wasm-bindgen because of its unique position in the ecosystem. For other binding generators that are only used by a couple of projects each, and for projects that write their wasm-host FFI by hand, the migration costs can't be amortized in the same way. That's not to say they shouldn't have been informed earlier and more deliberately about the looming ABI change. But I also don't see how the migration path for them could be significantly better than being told "in Rust 1.x the meaning of `extern "C"` on wasm32-unknown-unknown will change like this, please adjust your FFI code accordingly".

**Comment by bjorn3** [comment]
For Diplomat would it be possible to support both the old and new wasm ABI in the same Diplomat version provided that you had a way to detect which of the two was used for the wasm module? If so it should be possible to keep supporting the MSRV of ICU4X while also being compatible with a future rustc that switches to the new ABI by default, right?

**Comment by Manishearth** [comment]
> That's not to say they shouldn't have been informed earlier and more deliberately about the looming ABI change. But I also don't see how the migration path for them could be significantly better than being told "in Rust 1.x the meaning of `extern "C"` on wasm32-unknown-unknown will change like this, please adjust your FFI code accordingly".


I see a lot of potential routes. One would be `extern "C-wasm-legacy"`: being able to pick this for specific functions isn't too bad.  There's a lot you can do within the world of changing ABI annotations: and tools/users can even expose a cfg for this if they wish to support wider ranges of Rust versions.

A cfg flag would also help. It would let people generate JS bindings which can check which ABI a Rust module was built with by checking for the value of some const or something. This would effectively mean shipping two copies of the JS bindings and having a load-time switch but that's .... fine. It's not the best solution, but it's fine.

These are the two that I've thought of right now in the last day. I'm sure there are more.

I think it's rather presumptuous to assume there isn't a migration path when there simply was never an opportunity to discuss and ideate on this, because as you say, _we should have been informed earlier and more deliberately_. Not doing that means that we never got to explore the design space here. I'm sure there are reasons why `"C-wasm-legacy"` is a bad idea, but the whole thing is not worth discussing at this point.

Even _having prior warning about the timescale_ would have been quite helpful to decide on solutions. A lot of potential solutions are no longer open to us because of the short timescale.



@bjorn3 

> For Diplomat would it be possible to support both the old and new wasm ABI in the same Diplomat version provided that you had a way to detect which of the two was used for the wasm module? If so it should be possible to keep supporting the MSRV of ICU4X while also being compatible with a future rustc that switches to the new ABI by default, right?


... kind of, it's not ideal: the change needs to be in the generated JS, not Rust, but you can use the cfg to expose some flag function or something  and let the generated JS code, at wasm load time, decide which code to load for the API.

**Comment by Manishearth** [comment]
Overall I do think a lot of this could have been avoided were I to have engaged earlier to make sure I understood the migration plan, and initiated discussions when I discovered there wasn't really a migration plan. If we were having this discussion a year ago, making a will-be-removed-eventually `extern "C-wasm-legacy"` would have been an actually useful route forward here. It's not anymore, unless you want to wait another year or so to do the cutover.

But as I said, I do not think it was unreasonable for me to see the flag and expect the standard Rust process of providing multiple routes on stable for a long time before removing something, and thus assume I was already set.

I'm voicing this now so that (hopefully) the compiler team is more careful about this in the future.

**Comment by sffc** [comment]
Hi! I contribute to the Diplomat project along with @Manishearth.

Firstly, I am happy that y'all are proposing a proper Rust-WASM ABI. It is sorely needed.

I just wanted to respond to:

> wasm-bindgen as the one known user was fixed.

I work on software libraries (ICU/ICU4X) and language standards myself (I currently sit in ECMA TC39). I frequently find myself in situations where I want to "fix a bug", but there's some random client somewhere who comes out of the woodwork depending on the old behavior. My job is to proactively hunt down such clients and work with them on a migration strategy. This is consistent with the philosophy that the person who wants to break something holds the burden of making the breakage as non-disruptive as possible.

Diplomat, and by extension any downstream projects that use Diplomat, is and has been a user of the current Rust Wasm ABI since about 2021. We've regularly communicated with various folks in the WASM and Rust community, and we've given talks at local meetups about Diplomat. Everything Diplomat has been very much out in the open on GitHub and on crates.io. I'm not sure how we _could have_ been any more discoverable.

A postmortem should therefore look at why wasm-bindgen was the "only known user".

**Comment by hanna-kruppe** [comment]
@Manishearth I don't want to spill much more digital ink on what-ifs for the same reasons you outline. I want to apologize for coming across presumptuous when discussing technical challenges for better migration paths, but at the same time point out that *those discussions did happen*, just not with y'all involved and thus focused mostly on wasm-bindgen. Several of the possibilities you mention were brought up by others a long time ago but ultimately not pursued (e.g., stabilizing and eventually deprecating `-Zwasm-c-abi` was mentioned [the implementation PR](https://github.com/rust-lang/rust/pull/117919) and [compiler MCP](https://github.com/rust-lang/compiler-team/issues/703)), or even implemented unstably and reverted later ([`extern "wasm"` ABI](https://github.com/rust-lang/rust/issues/83788)). Of course, having more stakeholders involved earlier might have tilted the trade-offs towards a different outcome, or at least given other projects as much time to prepare and implement their side of the migration (as wasm-bindgen had). But it's not like the design space of what technical migration paths the Rust project could offer wasn't explored at all.

**Comment by Manishearth** [comment]
Someone already pointed me at the `extern "wasm"` discussion: it's not being discussed as a migration path, it's being discussed as a solution in and of itself, and it was dismissed as a bad solution ("we don't want to encourage the usage of unstable buggy stuff"). It's also tangled with multival. So yeah, I do think the outcome might have been different if there had been stakeholders putting forth their needs for migration.

**Comment by hanna-kruppe** [comment]
Migration was explicitly part of the [motivation when the "wasm" ABI was introduced](https://github.com/rust-lang/rust/pull/83763):

> Eventually remove the "wasm-bindgen compat hack". Once this multivalue
ABI is stable wasm-bindgen can switch to using it everywhere.
Afterwards the wasm32-unknown-unknown target can have its default ABI
updated to match C.

**Comment by Manishearth** [comment]
It wasn't a part of the discussion when it was removed. It was introduced four years ago, and the primary motivation there was to have a path to multival.

Anyway, not really worth rabbitholing on, but I basically do not see strong enough arguments against using it as a migration path, folks just didn't need one for wasm bindgen so it was deemed unnecessary.

**Comment by hanna-kruppe** [comment]
I just wanted to be clear that I don’t dismiss alternative migration paths because I personally can’t think of one (and I’m sorry for giving that impression), but because I believe alternatives were seriously explored (although under the mistaken assumption that only wasm-bindgen would have to migrate) and found lacking in ways that seem fairly fundamental to me (e.g., the reasons cited for removal of `extern "wasm"` are mostly not about wasm-bindgen). If you disagree with this characterization of past discussions/explorations, then yeah there’s no point in arguing further about this.

**Comment by RalfJung** [comment]
So it seems like the state of ABI in wasm is so bad that the ABI implemented by rustc didn't even stand out as being non-standard, and people are also used to reverse engineering the ABI and assuming they can rely on that as a stable promise? That's a pretty sad state of affairs, way worse than any other target I am aware of. :/

---

As before, I am not convinced by your claim that this is a breaking change in the usual sense: breaking some explicit stable promise. You have not provided further support for that claim. So I don't agree with the expectations you have set up here. It is more like people relying on assumptions about the Rust semantics that are not actually given and that are being broken by new optimizations (as you suggested yourself by bringing up Tree Borrows). There's little established process for that, and the usual "show a future-compat lint" is typically not an option.

That said, this is clearly a breaking change in that it breaks some users, and we should do our best to not let down our users even when they rely on what we think are unstable implementation details, and so far we didn't do a great job at involving those users in this discussion.

---

Regarding my earlier statement that...

> wasm-bindgen as the one known user was fixed.

I should not have stated that with the authority that I did. I was not actually involved in the discussions that led here, I'm just loosely involved in anything that involves fixing I-unsound ABI bugs, and I did some cleanup in rustc's ABI handling which found a whole bunch of bugs but also needed a special exception for the wasm ABI due to its broken nature.

I should instead let the people who drove the process speak for themselves. Though I am not even sure who that was... @bjorn3 and @alexcrichton ?

---

FWIW I think it's not too late to have a blog post, ideally written by someone who has all the context for this issue. However, I am not sure what concrete course of action the blog post should announce -- will we have a "wasm-legacy" ABI string to keep the old behavior around a bit longer or not? I'd love to see the annoying hack in rustc that currently keeps this working go away, but OTOH keeping it around a bit longer would not be the end of the world as long as there is a clear plan for removing it eventually.

A post-mono lint for non-scalar arguments passed across the "C" ABI on wasm32-unk-unk should be fairly easy to implement by adding some more logic in `compiler/rustc_monomorphize/src/mono_checks/abi_check.rs`.

**Comment by Manishearth** [comment]
> So it seems like the state of ABI in wasm is so bad that the ABI implemented by rustc didn't even stand out as being non-standard, and people are also used to reverse engineering the ABI and assuming they can rely on that as a stable promise? That's a pretty concerning state of affairs, way worse than any other architecture I am aware of. :/

Yeah pretty much. Wasm has always been a mess, it is becoming less of a mess (thank you all for working on that), but it is still a mess.

In this case, it is worth highlighting that [the standard](https://github.com/WebAssembly/tool-conventions/blob/main/BasicCABI.md) doesn't really connect up the ABI to what actually happens at the JS level, and while you can look at other standards for this, it's rather unclear exactly what's going on. It talks about stacks and then immediately starts talking about direct and indirect passing and function parameters. There really isn't any good documentation with that use case in mind: this is like if figuring out how to call Rust from C required you to understand register-passing first.

So while I do recall reading this document at some point, at the time the discrepancy between the document and observed behavior in Rust seemed to be a deficiency in the document, which did not seem to be written with my use case as a primary audience. It didn't occur to me that Rust would be nonstandard here, as I said, Rust and Wasm has been touted as this star-crossed pairing for ages, marketed as _the preferred_ way to do Wasm. It is not unreasonable to expect that Rust would be standard here, so much of the wasm effort has been driven by Rust people.

Basically, this space is quite underdocumented, the only documentation is rather terse standards which only provide part of the answer and need to be linked up with other standards to really figure things out. I don't really expect users to do anything _other_ than reverse engineer in this case.



> As before, I am not convinced by your claim that this is a breaking change in the usual sense: breaking some explicit stable promise. You have not provided further support for that claim.


I don't think this is a meaningful question to ask: there are no such documented promises at the moment, because Rust's stability policy has always been ["you should never have to fear upgrading to a new version of stable Rust"](https://doc.rust-lang.org/book/appendix-07-nightly-rust.html#stability-without-stagnation), with [explicit carve-outs for things like soundness](https://rust-lang.github.io/rfcs/1122-language-semver.html). It is not "we declare these things to be explicitly stable and everything else is whatever", it's the other way around.

And yes, this issue can be considered to be a soundness issue, but [the care required by that policy does not seem to have been put in](https://rust-lang.github.io/rfcs/1122-language-semver.html#soundness-changes). Even just future incompat linting about aggregates over extern-C-wasm would probably have helped identify problems!

Perhaps this points to a need for a newer stability policy that covers more things. The current stability policy [has a bunch of carve-outs for underspecified semantics](https://rust-lang.github.io/rfcs/1122-language-semver.html#underspecified-language-semantics) but they are very much not the full set: it doesn't even include `extern "Rust"` and `repr(Rust)`, which are widely agreed upon to be unspecified and liable to change. When it was written, there was not much thought given to the type of work that is ongoing today with specifying the semantics of Rust.


The `extern "Rust"` ABI is explicitly unstable (not by documented policy, but it is _well known_ that it is), as is `repr(Rust)`[^1]. I do not think it is unreasonable to expect the `extern "C"` ABI to be stable. I grant that it was buggy and needed fixing, but I don't think that was readily apparent, and as I have said before it feels like Rust has only ever given extremely positive signals about how perfect it is for Wasm and how the Rust and Wasm ecosystems are so easy to integrate. Expecting the simple FFI task of a function call to be stable is still not unreasonable in my book.

This isn't "relying on ABI" in the sense of relying on how register/stack stuff works: it's about how that translates to JS calls, _the primary purpose of writing `extern "C"` in Rust for Wasm_.



I really do think that one potential action item out of this discussion is that the compiler/lang team should publish a (non RFC, but living document) policy that lists both explicit stable promises, and explicit stable non-promises. From various discussions I do not think the current team's policy is aligned with what I knew the policy to be back when Rust 1.0 happened, and is not aligned with how the compiler/lang team thought back then, which is fine, this stuff sometimes evolves, but it would greatly help those stewarding large Rust codebases. I ask for this not just with my ICU4X hat on, but as someone doing tons of unsafe review at my company.

> I should instead let the people who drove the process speak for themselves.



Observation: the MCP was in 2023. Diplomat has had a kinda-experimental JS backend since 2021, but it did not implement the full ABI: its author found docs about the return value handling but not much more on the actual parameter ABI, and figured it out by looking at the code. At the time, we knew the implementation was buggy and needed work, but it was low priority to investigate it, and we planned to eventually find wasm experts to help us.

Diplomat's current rewritten JS backend was written in mid-2024, and ambiguousname did a lot more investigating on the ABI, fixing subtleties he discovered. With the new backend we had a lot more confidence about the wasm behavior due to that work being put in. 

So perhaps a thing to keep an eye out for is for multi-year processes like this, the concern of users and migration plans may need to be revisited to ensure users are not cropping up.

> I'd love to see the annoying hack in rustc that currently keeps this working go away, but OTOH keeping it around a bit longer would not be the end of the world as long as there is a clear plan for removing it eventually

I'd love that! It would be suboptimal to force Diplomat to not pass aggregates over FFI in wasm: there are ways to implement that behavior, but they are not great, and break Diplomat's model of "one common `extern "C"` layer for each language".




 [^1]: Which is what the previous issue was about. In that issue I was not arguing that Rust incorrectly made a breaking change: I recognized that `repr(packed)` without `repr(C)` was underdefined, I was just suggesting that I thought others may have assumed `repr(packed)`  implies `repr(C)`, and some more care needed to be taken to avoid breaking people.

**Comment by RalfJung** [comment]
> Perhaps this points to a need for a newer stability policy that covers more things. The current stability policy [has a bunch of carve-outs for underspecified semantics](https://rust-lang.github.io/rfcs/1122-language-semver.html#underspecified-language-semantics) but they are very much not the full set: it doesn't even include extern "Rust" and repr(Rust),

It kind does mention `repr`, though not ABI, and for some reason *after* the list:

"One obvious example is the layout of data structures, which is considered undefined unless they have a #[repr(C)] attribute."


> I really do think that one potential action item out of this discussion is that the compiler/lang team should publish a (non RFC, but living document) policy that lists both explicit stable promises, and explicit stable non-promises. From various discussions I do not think the current team's policy is aligned with what I knew the policy to be back when Rust 1.0 happened, and is not aligned with how the compiler/lang team thought back then, which is fine, this stuff sometimes evolves, but it would greatly help those stewarding large Rust codebases. I ask for this not just with my ICU4X hat on, but as someone doing tons of unsafe review at my company.

That sounds like a larger discussion to be had separately.

>  I grant that it was buggy and needed fixing, but I don't think that was readily apparent,

I think this is where your perception differs from mine (I cannot speak for the rest of t-compiler). As a total outsider to the wasm ecosystem, I assumed wasm would work roughly like the other architectures, in particular meaning there are well-defined ABI rules and people wouldn't think that reverse engineering the undocumented details of register-level argument passing is somehow normal.

---

Nominating for t-compiler to determine the path forward. I roughly see 3 options:
- Add a FCW for passing types across an `extern "C"` boundary on wasm32-unk-unk that are affected by this flag
- FCW + a blog post announcing our plans
- FCW + an `extern "wasm-legacy"` ABI that is explicitly planned to be removed again at some point in the future + blog post

**Comment by Manishearth** [comment]
> Add a FCW for passing types across an extern "C" boundary on wasm32-unk-unk that are affected by this flag

n.b. my current hope is to just ride it out for diplomat with explicit instructions for users to set appropriate flags, so I hope this FCW doesn't become an actual error; I do wish to have the ability to pass aggregates over wasm ffi in the long run.

An FCW that basically warns people that they are in a period where they should be careful about their compiler version + `-Zwasm-c-abi` pointing to some docs that explain things more, would probably be quite good. And then it gets removed eventually after the team judges sufficient time has passed.


I _may_ decide to try something like wasm-bingen's solution if I can manage. It's tempting, especially with the performance wins of avoiding allocation.

(Happy to help with the post, btw)

**Comment by Manishearth** [comment]
> I assumed wasm would work roughly like the other architectures, in particular meaning there are well-defined ABI rules and people wouldn't think that reverse engineering the undocumented details of register-level argument passing is somehow normal.

Yeah, it's too new, _and_ the primary compiler used for wasm is .... rustc, so.......

**Comment by RalfJung** [comment]
So far in my experience, wasm being new meant everything was extremely well-done as they learned from all the past mistakes. There's even a formal language spec :)

But I have only been following from afar.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/138601 implements the suggested lint.

> n.b. my current hope is to just ride it out for diplomat with explicit instructions for users to set appropriate flags, so I hope this FCW doesn't become an actual error; I do wish to have the ability to pass aggregates over wasm ffi in the long run.

I'm fine with just removing the FCW in the same PR that also does the ABI switch (without ever having it be an actual error), if that works for you.

**Comment by RalfJung** [comment]
> My understanding is that someone didn't realize they have to add some code in rustc's ABI handling whenever a new architecture is being added, and so the ABI is some undocumented artifact of what LLVM happens to do.

That doesn't seem to be entirely true, this was at least partially intentionally:
```
/// The purpose of this ABI is for matching the WebAssembly standard. This
/// intentionally diverges from the C ABI and is specifically crafted to take
/// advantage of LLVM's support of multiple returns in WebAssembly.
```
However, whoever wrote this clearly did not consider that the way rustc translates Rust types to LLVM types is not stable so we can't *ever* just use the "bare" LLVM ABI in this way.

Anyway at that point this ism only of historical interest.

**Comment by hanna-kruppe** [comment]
This comment from the comes from the introduction of `extern "wasm"` (https://github.com/rust-lang/rust/pull/83763), which was much later. That ABI was intended to be stable eventually, not necessarily identical to the "legacy" accident, and defined in terms of wasm concepts. So presumably if it had become stable then it would have had an LLVM-independent definition, with more deliberate lowering to LLVM IR added as necessary.

**Comment by alexcrichton** [comment]
This has gotten to be quite a long thread but I can try to comment on at least some parts from a wasm target maintainer perspective. IMO the main thing here is that the ABI was "broken" from the get-go (my mistake) and then by the time we discovered it (years later) I had run out of steam and couldn't champion the change to wasm-bindgen. This then started a multi-year process (5+ I think?) to change the ABI where no one person was ever solely dedicated to completing the migration. Naturally thoughout this time people have come and gone and progress has been made but it's inevitable that with such a large cast of characters and such a large amount of time that things fall through the cracks. When this started (which I think was 2020 or pre-2020, I forget), I'm relatively certain wasm-bindgen was the only thing in play, and I think that mentality stuck by default despite that not being true years later.

Regardless of this past though it seems like we're all aligned on (a) everyone here's operating in good-faith (yay!) and (b) the goal is to change the ABI still. I definitely agree the ABI should still change here myself, and on the topic of changing from a "good ABI" to a "not as good ABI" in terms of what the actual ABI is, this is another long-running issue on the wasm side and I think that discussion should happen on tool-conventions (I can point to issues if desired) rather than here. The goal here is to align with Clang and stop implicitly relying on Rust compiler internals.

Personally I'd think it best if we could get away with FCW + blog. Question for you though @Manishearth as that has pretty serious implications for Diplomat, are you ok with these? (sorry if this was already discussed)

* Users of rustc stable and Diplomat would receive warnings that cannot be addressed without switching to a nightly toolchain and passing `-Zwasm-c-abi=spec`.
* Users that switch from stable to nightly will be guaranteed to break again in the future due to `-Zwasm-c-abi` eventually going to get removed.
* Users that decide to pin to a rustc stable version instead of updating might be stuck for on the order of 3-6 months.

If you're ok with that, that seems like a reasonable plan to move forward to me, with a possible schedule looking like:

1. Land FCW + write blog post.
2. Wait for FCW to hit stable
3. Change `-Zwasm-c-abi` default in nightly once FCW is on stable
4. Wait for new default to hit stable
5. Remove FCW + `-Zwasm-c-abi`

That's going to be probably a longer transition period than envisioned (I think?). (1) here can happen here-and-now, but (3) is ~12 weeks out and (5) is probably minimum 24 weeks out, possibly more.

**Comment by Manishearth** [comment]
That seems good. The schedule is a bit fast according to ICU4X standards (we change MSRV slowly), but if the FCW has helpful advice for people in this situation, and is framed less like an FCW and more like "this is compiler-version dependent, take care, here's a link for how to fix", and the lint goes away with explicit `wasm-c-abi`, then that shouldn't matter much. 

I will discuss this with other ICU4X and Diplomat team members to ensure everyone is on the same page.

**Comment by workingjubilee** [comment]
> So far in my experience, wasm being new meant everything was extremely well-done as they learned from all the past mistakes. There's even a formal language spec :)
> 
> But I have only been following from afar.

They only learned from things people wrote down and made public.

The horrors of the hard lessons learned in, say, the evolution of the x86 Microsoft C++ ABI, seem to have been buried in Redmond.

**Comment by alexcrichton** [comment]
> I was thinking we'd remove the FCW the moment we change the default.

To me this depends on the transition plan. It sounds like Diplomat's plan is something along the lines of:

1. Rely on the legacy ABI for the time being.
2. Get current users to pin their toolchain or switch to Nightly with `-Zwasm-c-abi=legacy`
3. Eventually Rust switches the default where `-Zwasm-c-abi=spec` is the new default
4. Eventually this change reaches stable
5. Then Diplomat changes internal JS code and such to only work with this new stable compiler, updating MSRV. 
  * Users pinned to old toolchains "just update" now
  * Users using nightly with `-Zwasm-c-abi=legacy` will probably see corruption/crashes if they update Diplomat without removing the legacy flag, but the fix would then be to remove the legacy flag.

If the FCW is removed when the default is changed (which I take to also remove `-Zwasm-c-abi`) then there's a period of time between (3) and (4) where Diplomat is unsupported on all of stable/beta/nightly. Whether or not that's ok for Diplomat I'm not sure, but by delaying the removal of the FCW + `-Zwasm-c-abi` it provides a bit of a grace period for Diplomat to still be supported on one of stable/beta/nightly. 

I should definitely clarify though that to me I'm considering two knobs: one is the change of the default value of `-Zwasm-c-abi` and another is the removal of all support for the legacy abi. Removal of the FCW I would bundle with the removal of the legacy ABI.

> The schedule is a bit fast according to ICU4X standards

Oh I'm just throwing out ideas. I'm not actually in charge of timelines or anything. Personally I think it's valuable for everything to have a bounded amount of time, but extending a release or two is probably fine.

> the lint goes away with explicit `wasm-c-abi`

To clarify, and to be double-sure: this is not possible with a stable or beta compiler, only a nightly compiler. My thinking is there is no plan to move `-Zwasm-c-abi` to `-C`. It stays in the perma-unstable category of `-Z` meaning the only way to silence the warning (and insulate a project from future breakage) will be (a) pinning rustc to an older version or (b) switch to Nightly + `-Zwasm-c-abi=legacy`. 

I want to be extra careful to clarify this since most projects in the past I know of are hesitant to do both, but if that's ok for Diplomat it definitely makes the compiler's life easier

**Comment by RalfJung** [comment]
> If the FCW is removed when the default is changed (which I take to also remove -Zwasm-c-abi) 

No, the flag can stay around a bit longer once the default is changed. Just the FCW is gone then. (Or, impossible to hit on stable. It might still exist and trigger for people that set `-Zwasm-c-abi` I guess 🤷 .)

**Comment by alexcrichton** [comment]
Ah ok, I think you and I are thinking the same thing then, so sounds good!

In some sense the FCW "will break" warning is correct with `-Zwasm-c-abi=legacy` because the plan is to remove the flag entirely which would break builds using that. Anyway I digress...

**Comment by Manishearth** [comment]
> To clarify, and to be double-sure: this is not possible with a stable or beta compiler, only a nightly compiler

Yep. But one way to deal with the transition is, during it, to explicitly pass a wasm-c-abi flag on nightly. It's not _great_ to have to use nightly for this, but for building `.wasm` files it's fine.

**Comment by alexcrichton** [comment]
[Up above](https://github.com/rust-lang/rust/issues/122532#issuecomment-2728428935) @RalfJung added I-compiler-nominated, so in an effort to expedite team discussion of this I'm going to try to summarize the current status with respect to the compiler team. The main points are:

* Today `wasm32-unknown-unknown` still has a broken non-spec-conforming definition of the C ABI (this has been true for a long time)
* The ABI can be fixed by passing `-Zwasm-c-abi=spec`  (and the broken ABI is `-Zwasm-c-abi=legacy`)
* One major known user this would break is `wasm-bindgen`. Through a series of PRs over the years this is now "done" and is ready for the the ABI to change at any time.
* @Manishearth has brought up in this thread that another user, Diplomat, was not ready for this change and would be broken from the removal of `-Zwasm-c-abi=legacy` (which I believe was planned to happen "soon" but there's no current PR for this)
* Thus the current proposal of how to move forward and reach the end state of removing `-Zwasm-c-abi=legacy`
  * Add a [future-compat warning on usage of `-Zwasm-c-abi=legacy`](https://github.com/rust-lang/rust/pull/138601) where the meaning is changing. This should not warn on wasm-bindgen for example, only on ABI signatures whose definition is going to change when `-Zwasm-c-abi=spec` becomes the default.
  * When that change lands publish a blog post about `-Zwasm-c-abi` and explain what's going on. The "fix" is (a) stop using aggregates in your ABI, (b) use a nightly compiler and pass `-Zwasm-c-abi=spec`, or (c) pin to an older rustc for awhile
  * After the future-compat warning change bakes for a bit, change the default on Nightly from `-Zwasm-c-abi=legacy` to `-Zwasm-c-abi=spec`.
  * After the changed default bakes for awhile, remove `-Zwasm-c-abi` entirely and all support for `legacy`, including the future-compat warning.

My understanding is that this plan works for Diplomat and the hope is that this works for other users as well if they come out of the woodwork. Also this is my best understanding of the conclusions of this thread, if I got something wrong here someone please let me know.

**Comment by Manishearth** [comment]
The current plan that Diplomat and ICU4X are coalescing around is:

 - We shall land a patch in Diplomat around the time the default switches over. Until then, we basically ignore the FCW.
 - ICU4X 2.0, and whatever Diplomat version comes with that, will have legacy bindings. Unfortunate but fine. We hope for a release in the next few weeks. Users who need to use this on wasm may need to use an older stable compiler, or a nightly with the `legacy` flag.
 - We may have some kind of check that ensures the correct ABI was used, loudly erroring if you attempt to run the wasm bindings with a spec ABI `.wasm` file
 - ICU4X 2.1, and a semver incompatible Diplomat version, will have spec bindings. Users who need to use this on wasm should use older compilers, or nightlies with the `spec` flag. We may still have a legacy check if needed.


This means that the plan for an FCW and eventual switchover in the near future is probably acceptable to us. Will keep you posted if other needs come up.

**Comment by RalfJung** [comment]
@alexcrichton 

> The "fix" is (a) stop using aggregates in your ABI, (b) use a nightly compiler and pass -Zwasm-c-abi=spec, or (c) pin to an older rustc for awhile

`-Zwasm-c-abi=spec` is a "fix" but does opt-in to the ABI change, so the moment you do that you have to update the other side of this ABI as well.

So another "fix" is to set `-Zwasm-c-abi=legacy`. That gives you control over when the switch is performed -- it's when you toggle that to `spec`.

@Manishearth 

> ICU4X 2.1, and a semver incompatible Diplomat version, will have spec bindings. Users who need to use this on wasm should use older compilers, or nightlies with the spec flag. We may still have a legacy check if needed.

I am confused here, how would an older compiler help when using spec ABI bindinings?

**Comment by Manishearth** [comment]
typo, sorry!

**Comment by Manishearth** [comment]
> * remove `-Z` flag and finally remove the legacy wasm implementation for Rust 1.91

Might be nice to keep the `-Zwasm-c-abi=spec` as a no-longer-necessary flag that does nothing for a while after, just so that people using this flag don't have to do rust version detection. Not necessary, but nice to have.

**Comment by RalfJung** [comment]
> Might be nice to keep the -Zwasm-c-abi=spec as a no-longer-necessary flag that does nothing for a while after, just so that people using this flag don't have to do rust version detection. Not necessary, but nice to have.

As a nightly-only flag I assume people using it would anyway track latest nightly fairly closely?

**Comment by Manishearth** [comment]
> As a nightly-only flag I assume people using it would anyway track latest nightly fairly closely?

There can be other constraints on the nightly version you use. Not relevant for this particular thing, but e.g. ICU4X pins some nightlies in its tests so that we can track cross-language LTO codesize.

**Comment by Manishearth** [comment]
> In this case, it is worth highlighting that [the standard](https://github.com/WebAssembly/tool-conventions/blob/main/BasicCABI.md) doesn't really connect up the ABI to what actually happens at the JS level, and while you can look at other standards for this, it's rather unclear exactly what's going on. It talks about stacks and then immediately starts talking about direct and indirect passing and function parameters. There really isn't any good documentation with that use case in mind: this is like if figuring out how to call Rust from C required you to understand register-passing first.

Just to give another example of this: The multivalue extension has been talked about for ages. It is implemented everywhere[^1]. Yet, nowhere, do I see any mention of what this means in JS. I had to reverse engineer it, turns out that it returns an array of the scalars splatted out, which is reasonable, but it's really hard to find any docs that tell you how this work. Not just official docs, _any docs whatsoever_.

An observation about the wasm ecosystem from a relative outsider: Wasm was initially marketed as this thing for running fast native code from JS, but the focus has overall shifted to a more general usage as a way to safely embed native code in many kinds of applications that have nothing to do with JS. It feels like this has ended up with the JS-wasm use case getting less attention over time, leading to an overall lack of docs. I think things like this issue are also a knock on effect: where folks were trying to fix the C interop problem but did not really consider there to be any users of JS-wasm outside of wasm-bindgen, despite JS-wasm being the original primary purpose of wasm. I had to struggle to convince people that it was possible to use the current ABI because people were convinced it was unusably broken (it was broken, just not unusably so for JS-wasm).


 [^1]: For some reason I couldn't convince rustc to emit it, I was using `RUSTFLAGS="-C target-feature=+multivalue -Zwasm-c-abi=spec" cargo +nightly rustc -Zbuild-std=panic_abort,std  --target wasm32-unknown-unknown   --crate-type cdylib  --release`, but everything else points to it being supported, and I'm probably just doing something incorrect here.

**Comment by apiraino** [comment]
T-compiler visited this issue during the triage meeting ([on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bwekly.5D.202025-03-20/near/506998770)[#t-compiler/meetings > &#91;wekly&#93; 2025-03-20 @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bwekly.5D.202025-03-20/near/506998770)).

First of all thanks @RalfJung and @alexcrichton  for the great effort to summarize the main points. We basically have nothing to add and find the plan sound (although understandadly a bit of a burden for the parties involved).

We take from [this comment](https://github.com/rust-lang/rust/issues/122532#issuecomment-2727829699) an actionable for us to be more transparent and document better stability promises (or lack thereof).

@rustbot label -I-compiler-nominated

**Comment by alexcrichton** [comment]
If no one else wants to volunteer to write a blog post I'm happy to draft that up in the next few days with a plan to merge the blog post around the time #138601 lands.

**Comment by Manishearth** [comment]
@alexcrichton I can't take point on this but I can probably help review it and write a paragraph or two if you share a hackmd with me when you start.

**Comment by alexcrichton** [comment]
I've written up an [initial draft](https://github.com/rust-lang/blog.rust-lang.org/pull/1531)
